ouroboros-consensus-protocol/src/ouroboros-consensus-protocol/Ouroboros/Consensus/Protocol/Praos/AgentClient.hs

```diff
@@ -0,0 +1,183 @@
+      )
+  => AgentCrypto c where
+        type ACrypto c :: Type
+
+instance AgentCrypto StandardCrypto where
+  type ACrypto StandardCrypto = Agent.StandardCrypto
+
+convertOCert :: ( AgentCrypto c
+                )
+              => Agent.OCert (ACrypto c) -> OCert.OCert c
+convertOCert oca =
+  OCert.OCert
+    { OCert.ocertVkHot = Agent.ocertVkHot oca
+    , OCert.ocertN = Agent.ocertN oca
+    , OCert.ocertKESPeriod = OCert.KESPeriod (Agent.unKESPeriod $ Agent.ocertKESPeriod oca)
+    , OCert.ocertSigma = coerce (Agent.ocertSigma oca)
+    }
+
+class (MonadFail m, Show (Addr m)) => MonadKESAgent m where
+  type FD m :: Type
+  type Addr m :: Type
+  withAgentContext :: (Snocket m (FD m) (Addr m) -> m a) -> m a
+  makeRawBearer :: MakeRawBearer m (FD m)
+  makeAddress :: Proxy m -> FilePath -> Addr m
+
+instance MonadKESAgent IO where
+  type FD IO = Socket
+  type Addr IO = SockAddr
+  withAgentContext inner =
+    withIOManager $ \ioManager ->
+      inner (socketSnocket ioManager)
+  makeRawBearer = makeSocketRawBearer
+  makeAddress _ = SockAddrUnix
+
+instance MonadKESAgent (IOSim s) where
+  type FD (IOSim s) = SimSnocket.FD (IOSim s) (TestAddress FilePath)
+  type Addr (IOSim s) = TestAddress FilePath
+  withAgentContext inner = do
+    SimSnocket.withSnocket
+      nullTracer
+      (toBearerInfo $ absNoAttenuation {abiConnectionDelay = SmallDelay})
+      mempty
+      $ \snocket _observe -> inner snocket
+  makeRawBearer = SimSnocket.makeFDRawBearer nullTracer
+  makeAddress _ = TestAddress
+
+instance SimSnocket.GlobalAddressScheme FilePath where
+  getAddressType = const SimSnocket.IPv4Address
+  ephemeralAddress _ty num = TestAddress $ "simSnocket_" <> show num
+
+runKESAgentClient :: forall m c.
+                     ( KESAgentContext c m
+                     )
+                  => Tracer m KESAgentClientTrace
+                  -> FilePath
+                  -> (OCert.OCert c -> SignKeyKES (KES c) -> Word -> m ())
+                  -> m ()
+runKESAgentClient tracer path handleKey = do
+  withAgentContext $ \snocket -> do
+    forever $ do
+      Agent.runServiceClient
+        (Proxy @(ACrypto c))
+        makeRawBearer
+        (Agent.ServiceClientOptions
+          { Agent.serviceClientSnocket = snocket
+          , Agent.serviceClientAddress = makeAddress (Proxy @m) path
+          } :: Agent.ServiceClientOptions m (FD m) (Addr m)
+        )
+        (\(Agent.TaggedBundle mBundle _) -> do
+            case mBundle of
+              Just (Agent.Bundle skpRef ocert) -> do
+                -- We take ownership of the key, so we acquire one extra reference,
+                -- preventing the key from being discarded after `handleKey`
+                -- finishes.
+                _ <- acquireCRef skpRef
+                withCRefValue skpRef $ \(SignKeyWithPeriodKES sk p) ->
+                  handleKey (convertOCert ocert) sk p
+                return Agent.RecvOK
+              _ -> undefined -- TODO
```

Comment:
What to do with this?

Reply:
Ah, right; that's the case where the agent sends a "key drop" message. I'll take care of it.